File: app/core/smart_reframe/renderer.py
"""
FFmpeg-based rendering of reframed videos.

This module converts crop plans into FFmpeg commands
to produce the final reframed video files.
"""

import json
import logging
import subprocess
import tempfile
from pathlib import Path
from typing import Optional

from app.core.smart_reframe.models import (
    AspectRatio,
    CropPlan,
    CropWindow,
    ShotCropPlan,
)
from app.core.smart_reframe.config import IntelligentCropConfig
from app.core.smart_reframe.crop_planner import interpolate_crop_window

logger = logging.getLogger(__name__)


class Renderer:
    """
    Render reframed videos using FFmpeg.
    """

    def __init__(self, config: IntelligentCropConfig):
        self.config = config

    def render(
        self,
        input_path: str,
        crop_plan: CropPlan,
        output_prefix: str,
        output_resolution: Optional[tuple[int, int]] = None,
    ) -> dict[str, str]:
        """
        Render reframed videos for all target aspect ratios.

        Args:
            input_path: Path to source video.
            crop_plan: Computed crop plan.
            output_prefix: Prefix for output file names.
            output_resolution: Optional (width, height) for output.
                              If not provided, uses crop dimensions.

        Returns:
            Dict mapping aspect ratio string to output path.
        """
        output_paths = {}

        for aspect_ratio in crop_plan.target_aspect_ratios:
            # Collect crop plans for this aspect ratio
            aspect_crop_plans = [
                cp
                for cp in crop_plan.shot_crop_plans
                if cp.aspect_ratio == aspect_ratio
            ]

            if not aspect_crop_plans:
                logger.warning(f"No crop plans for aspect ratio {aspect_ratio}")
                continue

            # Generate output path
            aspect_str = f"{aspect_ratio.width}x{aspect_ratio.height}"
            output_path = f"{output_prefix}_{aspect_str}.mp4"

            # Determine rendering strategy
            if self._is_static_crop(aspect_crop_plans):
                self._render_static_crop(
                    input_path,
                    aspect_crop_plans,
                    output_path,
                    output_resolution,
                    crop_plan,
                )
            else:
                self._render_dynamic_crop(
                    input_path,
                    aspect_crop_plans,
                    output_path,
                    output_resolution,
                    crop_plan,
                )

            output_paths[str(aspect_ratio)] = output_path
            logger.info(f"Rendered {output_path}")

        return output_paths

    def _get_time_range(self, crop_plan: CropPlan) -> tuple[float, float]:
        """
        Get the overall time range from the crop plan's shots.
        """
        if not crop_plan.shots:
            return 0.0, crop_plan.video.duration

        start_time = min(shot.start_time for shot in crop_plan.shots)
        end_time = max(shot.end_time for shot in crop_plan.shots)
        return start_time, end_time

    def _is_static_crop(self, crop_plans: list[ShotCropPlan]) -> bool:
        """
        Check if all crop windows are essentially static.

        Returns True if a simple static crop can be used.
        """
        all_windows = []
        for cp in crop_plans:
            all_windows.extend(cp.crop_windows)

        if len(all_windows) <= 1:
            return True

        # Check variance in crop windows
        x_vals = [w.x for w in all_windows]
        y_vals = [w.y for w in all_windows]
        w_vals = [w.width for w in all_windows]

        x_range = max(x_vals) - min(x_vals)
        y_range = max(y_vals) - min(y_vals)
        w_range = max(w_vals) - min(w_vals)

        avg_width = sum(w_vals) / len(w_vals)

        # Consider static if movement is less than 5% of width
        threshold = avg_width * 0.05
        return x_range < threshold and y_range < threshold and w_range < threshold

    def _render_static_crop(
        self,
        input_path: str,
        crop_plans: list[ShotCropPlan],
        output_path: str,
        output_resolution: Optional[tuple[int, int]],
        full_plan: CropPlan,
    ):
        """
        Render using a single static crop (faster).
        """
        # Use the median crop window
        all_windows = []
        for cp in crop_plans:
            all_windows.extend(cp.crop_windows)

        if not all_windows:
            raise ValueError("No crop windows to render")

        # Compute median crop
        x = int(sorted(w.x for w in all_windows)[len(all_windows) // 2])
        y = int(sorted(w.y for w in all_windows)[len(all_windows) // 2])
        width = int(sorted(w.width for w in all_windows)[len(all_windows) // 2])
        height = int(sorted(w.height for w in all_windows)[len(all_windows) // 2])

        # Get time range from shots
        start_time, end_time = self._get_time_range(full_plan)

        # Build FFmpeg command
        vf_filters = [f"crop={width}:{height}:{x}:{y}"]

        if output_resolution:
            vf_filters.append(f"scale={output_resolution[0]}:{output_resolution[1]}")
        else:
            # Scale to even dimensions (required by many codecs)
            vf_filters.append("scale=trunc(iw/2)*2:trunc(ih/2)*2")

        vf = ",".join(vf_filters)

        cmd = [
            "ffmpeg",
            "-y",
            "-ss", f"{start_time:.3f}",
            "-i", input_path,
            "-t", f"{end_time - start_time:.3f}",
            "-vf", vf,
            "-c:v", "libx264",
            "-preset", self.config.render_preset,
            "-crf", str(self.config.render_crf),
            "-c:a", "aac",
            "-b:a", "128k",
            output_path,
        ]

        self._run_ffmpeg(cmd)

    def _render_dynamic_crop(
        self,
        input_path: str,
        crop_plans: list[ShotCropPlan],
        output_path: str,
        output_resolution: Optional[tuple[int, int]],
        full_plan: CropPlan,
    ):
        """
        Render using dynamic crop expressions or segment concatenation.

        This handles time-varying crops by generating FFmpeg filter
        expressions or splitting into segments.
        """
        # Sort shots by time
        shots_sorted = sorted(full_plan.shots, key=lambda s: s.start_time)

        # Check if we can use expression-based cropping
        # For simplicity, we use segment-based approach
        self._render_segments(
            input_path,
            crop_plans,
            output_path,
            output_resolution,
            full_plan,
        )

    def _render_segments(
        self,
        input_path: str,
        crop_plans: list[ShotCropPlan],
        output_path: str,
        output_resolution: Optional[tuple[int, int]],
        full_plan: CropPlan,
    ):
        """
        Render by splitting into segments with constant crops.
        """
        # Merge all crop windows sorted by time
        all_windows: list[tuple[float, CropWindow]] = []
        for cp in crop_plans:
            for cw in cp.crop_windows:
                all_windows.append((cw.time, cw))

        all_windows.sort(key=lambda x: x[0])

        if not all_windows:
            raise ValueError("No crop windows to render")

        # Group into segments with similar crops
        segments = self._group_segments(all_windows)

        if len(segments) == 1:
            # Single segment - use direct crop
            seg = segments[0]
            self._render_single_segment(
                input_path,
                seg["start"],
                seg["end"],
                seg["crop"],
                output_path,
                output_resolution,
            )
        else:
            # Multiple segments - concatenate
            self._render_and_concat_segments(
                input_path,
                segments,
                output_path,
                output_resolution,
            )

    def _group_segments(
        self,
        windows: list[tuple[float, CropWindow]],
        tolerance: float = 0.1,
    ) -> list[dict]:
        """
        Group crop windows into segments with similar crops.
        """
        if not windows:
            return []

        segments = []
        current_start = windows[0][0]
        current_crop = windows[0][1]
        prev_time = windows[0][0]

        for time, crop in windows[1:]:
            # Check if crop has changed significantly
            if self._crops_differ(current_crop, crop, tolerance):
                # End current segment
                segments.append(
                    {
                        "start": current_start,
                        "end": time,
                        "crop": current_crop,
                    }
                )
                current_start = time
                current_crop = crop

            prev_time = time

        # Add final segment
        segments.append(
            {
                "start": current_start,
                "end": prev_time + 1.0,  # Extend slightly
                "crop": current_crop,
            }
        )

        return segments

    def _crops_differ(
        self,
        crop1: CropWindow,
        crop2: CropWindow,
        tolerance: float,
    ) -> bool:
        """Check if two crops are significantly different."""
        threshold = crop1.width * tolerance
        return (
            abs(crop1.x - crop2.x) > threshold
            or abs(crop1.y - crop2.y) > threshold
            or abs(crop1.width - crop2.width) > threshold
        )

    def _render_single_segment(
        self,
        input_path: str,
        start: float,
        end: float,
        crop: CropWindow,
        output_path: str,
        output_resolution: Optional[tuple[int, int]],
    ):
        """Render a single segment with constant crop."""
        duration = end - start

        vf_filters = [f"crop={crop.width}:{crop.height}:{crop.x}:{crop.y}"]

        if output_resolution:
            vf_filters.append(f"scale={output_resolution[0]}:{output_resolution[1]}")
        else:
            vf_filters.append("scale=trunc(iw/2)*2:trunc(ih/2)*2")

        vf = ",".join(vf_filters)

        cmd = [
            "ffmpeg",
            "-y",
            "-ss", f"{start:.3f}",
            "-i", input_path,
            "-t", f"{duration:.3f}",
            "-vf", vf,
            "-c:v", "libx264",
            "-preset", self.config.render_preset,
            "-crf", str(self.config.render_crf),
            "-c:a", "aac",
            "-b:a", "128k",
            output_path,
        ]

        self._run_ffmpeg(cmd)

    def _render_and_concat_segments(
        self,
        input_path: str,
        segments: list[dict],
        output_path: str,
        output_resolution: Optional[tuple[int, int]],
    ):
        """Render segments and concatenate them."""
        with tempfile.TemporaryDirectory() as tmpdir:
            segment_files = []

            for i, seg in enumerate(segments):
                seg_path = Path(tmpdir) / f"segment_{i:04d}.mp4"

                self._render_single_segment(
                    input_path,
                    seg["start"],
                    seg["end"],
                    seg["crop"],
                    str(seg_path),
                    output_resolution,
                )

                segment_files.append(seg_path)

            # Create concat list
            concat_list = Path(tmpdir) / "concat.txt"
            with open(concat_list, "w") as f:
                for seg_path in segment_files:
                    f.write(f"file '{seg_path}'\n")

            # Concatenate
            cmd = [
                "ffmpeg",
                "-y",
                "-f",
                "concat",
                "-safe",
                "0",
                "-i",
                str(concat_list),
                "-c",
                "copy",
                output_path,
            ]

            self._run_ffmpeg(cmd)

    def _run_ffmpeg(self, cmd: list[str]):
        """Run an FFmpeg command with improved error handling."""
        from app.core.utils.ffmpeg import run_ffmpeg
        
        # Note: run_ffmpeg will modify cmd to add -loglevel, so log after calling
        logger.debug(f"Running FFmpeg command (log level will be added automatically)")

        try:
            result = run_ffmpeg(
                cmd,
                suppress_warnings=True,
                log_level="error",  # Suppress warnings, keep errors
                check=True,
            )
            if result.stdout:
                logger.debug(f"FFmpeg stdout: {result.stdout}")
        except RuntimeError as e:
            # Re-raise as-is (already formatted)
            raise
        except Exception as e:
            logger.error(f"FFmpeg failed: {e}")
            raise RuntimeError(f"FFmpeg rendering failed: {e}") from e


def render_with_letterbox(
    input_path: str,
    crop_window: CropWindow,
    output_path: str,
    target_resolution: tuple[int, int],
    blur_sigma: float = 30.0,
    config: Optional[IntelligentCropConfig] = None,
):
    """
    Render with blurred letterbox background instead of black bars.

    Args:
        input_path: Path to source video.
        crop_window: Crop window to use.
        output_path: Path for output video.
        target_resolution: (width, height) of output.
        blur_sigma: Blur strength for background.
        config: Optional configuration.
    """
    if config is None:
        config = IntelligentCropConfig()

    out_w, out_h = target_resolution
    crop = crop_window

    # Build complex filter for blurred background + sharp foreground
    # 1. Scale input to output size with blur for background
    # 2. Crop and scale for foreground
    # 3. Overlay foreground on background
    vf = (
        f"split=2[bg][fg];"
        f"[bg]scale={out_w}:{out_h}:force_original_aspect_ratio=increase,"
        f"crop={out_w}:{out_h},gblur=sigma={blur_sigma}[blurred];"
        f"[fg]crop={crop.width}:{crop.height}:{crop.x}:{crop.y},"
        f"scale={out_w}:{out_h}:force_original_aspect_ratio=decrease[cropped];"
        f"[blurred][cropped]overlay=(W-w)/2:(H-h)/2"
    )

    cmd = [
        "ffmpeg",
        "-y",
        "-i",
        input_path,
        "-vf",
        vf,
        "-c:v",
        "libx264",
        "-preset",
        config.render_preset,
        "-crf",
        str(config.render_crf),
        "-c:a",
        "aac",
        "-b:a",
        "128k",
        output_path,
    ]

    logger.debug(f"Running letterbox render: {' '.join(cmd)}")

    from app.core.utils.ffmpeg import run_ffmpeg
    
    try:
        run_ffmpeg(cmd, suppress_warnings=True, log_level="error", check=True)
    except RuntimeError as e:
        logger.error(f"FFmpeg letterbox render failed: {e}")
        raise

File: backend/crates/vclip-media/src/intelligent/renderer.rs
//! FFmpeg-based rendering of intelligent cropped videos.
//!
//! Converts crop windows into FFmpeg commands for final rendering.

use super::config::IntelligentCropConfig;
use super::crop_planner::is_static_crop;
use super::models::CropWindow;
use super::output_format::{PORTRAIT_WIDTH, PORTRAIT_HEIGHT};
use crate::error::{MediaError, MediaResult};
use std::path::Path;
use std::process::Stdio;
use tempfile::TempDir;
use tokio::process::Command;
use tracing::{debug, info};

/// Renderer for intelligent cropped videos.
pub struct IntelligentRenderer {
    config: IntelligentCropConfig,
}

impl IntelligentRenderer {
    /// Create a new renderer.
    pub fn new(config: IntelligentCropConfig) -> Self {
        Self { config }
    }

    /// Render a video with intelligent cropping.
    ///
    /// # Arguments
    /// * `input` - Input video path
    /// * `output` - Output video path
    /// * `crop_windows` - Computed crop windows
    /// * `start_time` - Start time in seconds
    /// * `duration` - Duration in seconds
    pub async fn render<P: AsRef<Path>>(
        &self,
        input: P,
        output: P,
        crop_windows: &[CropWindow],
        start_time: f64,
        duration: f64,
    ) -> MediaResult<()> {
        let input = input.as_ref();
        let output = output.as_ref();

        if crop_windows.is_empty() {
            return Err(MediaError::InvalidVideo("No crop windows to render".to_string()));
        }

        // Determine rendering strategy
        if is_static_crop(crop_windows) {
            debug!("Using static crop rendering");
            self.render_static_crop(input, output, crop_windows, start_time, duration)
                .await
        } else {
            debug!("Using dynamic crop rendering");
            self.render_dynamic_crop(input, output, crop_windows, start_time, duration)
                .await
        }
    }

    /// Render using a single static crop (faster).
    async fn render_static_crop<P: AsRef<Path>>(
        &self,
        input: P,
        output: P,
        crop_windows: &[CropWindow],
        start_time: f64,
        duration: f64,
    ) -> MediaResult<()> {
        let input = input.as_ref();
        let output = output.as_ref();

        // Use median crop window
        let crop = self.compute_median_crop(crop_windows);

        // Build FFmpeg filter
        // Crop to 9:16 region, then scale to exact 1080×1920 portrait output
        let vf = format!(
            "crop={}:{}:{}:{},scale={}:{}:flags=lanczos,setsar=1",
            crop.width, crop.height, crop.x, crop.y,
            PORTRAIT_WIDTH, PORTRAIT_HEIGHT
        );

        let mut cmd = Command::new("ffmpeg");
        cmd.args([
            "-y",
            "-ss",
            &format!("{:.3}", start_time),
            "-i",
            input.to_str().unwrap_or(""),
            "-t",
            &format!("{:.3}", duration),
            "-vf",
            &vf,
            "-c:v",
            "libx264",
            "-preset",
            &self.config.render_preset,
            "-crf",
            &self.config.render_crf.to_string(),
            "-c:a",
            "aac",
            "-b:a",
            "128k",
            "-movflags",
            "+faststart",
            output.to_str().unwrap_or(""),
        ])
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

        debug!("Running FFmpeg static crop: {:?}", cmd);

        let output_result = cmd.output().await.map_err(|e| {
            MediaError::ffmpeg_failed(format!("Failed to run FFmpeg: {}", e), None, None)
        })?;

        if !output_result.status.success() {
            let stderr = String::from_utf8_lossy(&output_result.stderr);
            return Err(MediaError::ffmpeg_failed(
                "FFmpeg failed",
                Some(stderr.to_string()),
                output_result.status.code(),
            ));
        }

        Ok(())
    }

    /// Render with dynamic cropping using continuous filter graph.
    ///
    /// **Key fix for PTS discontinuity (garbled flash)**: Uses a single-pass
    /// filter graph with sendcmd to update crop parameters dynamically,
    /// instead of rendering segments separately and concatenating.
    async fn render_dynamic_crop<P: AsRef<Path>>(
        &self,
        input: P,
        output: P,
        crop_windows: &[CropWindow],
        start_time: f64,
        duration: f64,
    ) -> MediaResult<()> {
        let input = input.as_ref();
        let output = output.as_ref();

        // Group windows into segments with similar crops
        let segments = self.group_segments(crop_windows);

        if segments.len() == 1 {
            // Single segment - use static crop
            return self
                .render_single_segment(
                    input,
                    output,
                    &segments[0],
                    start_time,
                )
                .await;
        }

        info!(
            "Continuous dynamic crop: {} segments over {:.2}s",
            segments.len(),
            duration
        );

        // Use continuous rendering to avoid PTS discontinuities
        self.render_continuous_dynamic(input, output, crop_windows, &segments, start_time, duration)
            .await
    }

    /// Render using continuous filter graph (no segment concatenation).
    ///
    /// This method uses sendcmd to dynamically update crop parameters,
    /// keeping timestamps continuous and avoiding decode errors.
    async fn render_continuous_dynamic<P: AsRef<Path>>(
        &self,
        input: P,
        output: P,
        crop_windows: &[CropWindow],
        segments: &[CropSegment],
        start_time: f64,
        duration: f64,
    ) -> MediaResult<()> {
        let input = input.as_ref();
        let output = output.as_ref();

        if segments.is_empty() || crop_windows.is_empty() {
            return Err(MediaError::InvalidVideo("No crop segments provided".to_string()));
        }

        // Build sendcmd script for crop updates
        let sendcmd_script = self.build_sendcmd_script(segments);
        let initial = &segments[0].crop;

        // Build filter graph with sendcmd for dynamic crop updates
        // Key elements:
        // 1. setsar=1 - Normalize pixel aspect ratio
        // 2. setpts=PTS-STARTPTS - Reset timestamps to avoid discontinuities
        // 3. sendcmd - Dynamically update crop parameters
        // 4. format=yuv420p - Ensure compatible pixel format
        // Build filter graph with sendcmd for dynamic crop updates
        // Final scale ensures exact 1080×1920 portrait output
        let filter_complex = format!(
            "[0:v]setsar=1,setpts=PTS-STARTPTS,format=yuv420p,\
             sendcmd=f='{script}',\
             crop@dyncrop=w={w}:h={h}:x={x}:y={y}:exact=1,\
             scale={out_w}:{out_h}:flags=lanczos,setsar=1[vout]",
            script = sendcmd_script,
            w = initial.width,
            h = initial.height,
            x = initial.x,
            y = initial.y,
            out_w = PORTRAIT_WIDTH,
            out_h = PORTRAIT_HEIGHT,
        );

        debug!("Continuous crop filter:\n{}", filter_complex);

        let mut cmd = Command::new("ffmpeg");
        cmd.args([
            "-y",
            "-ss", &format!("{:.3}", start_time),
            "-i", input.to_str().unwrap_or(""),
            "-t", &format!("{:.3}", duration),
            "-filter_complex", &filter_complex,
            "-map", "[vout]",
            "-map", "0:a?",
            // Video encoding
            "-c:v", "libx264",
            "-preset", &self.config.render_preset,
            "-crf", &self.config.render_crf.to_string(),
            "-pix_fmt", "yuv420p",
            // Constant frame rate to prevent timing issues
            "-vsync", "cfr",
            "-video_track_timescale", "90000",
            // Audio with resample to handle any discontinuities
            "-c:a", "aac",
            "-b:a", "128k",
            "-af", "aresample=async=1:first_pts=0",
            // Output
            "-movflags", "+faststart",
            output.to_str().unwrap_or(""),
        ])
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

        let output_result = cmd.output().await.map_err(|e| {
            MediaError::ffmpeg_failed(format!("Failed to run FFmpeg: {}", e), None, None)
        })?;

        if !output_result.status.success() {
            let stderr = String::from_utf8_lossy(&output_result.stderr);
            
            // Fall back to legacy concat if sendcmd not supported
            if stderr.contains("sendcmd") || stderr.contains("Unknown filter") {
                debug!("sendcmd filter not available, falling back to segment concat");
                return self.render_and_concat_segments_fixed(input, output, segments, start_time).await;
            }
            
            return Err(MediaError::ffmpeg_failed(
                "Continuous crop render failed",
                Some(stderr.to_string()),
                output_result.status.code(),
            ));
        }

        Ok(())
    }

    /// Build sendcmd script for dynamic crop updates.
    fn build_sendcmd_script(&self, segments: &[CropSegment]) -> String {
        segments
            .iter()
            .map(|seg| {
                format!(
                    "{:.3} [enter] crop@dyncrop w {}, crop@dyncrop h {}, crop@dyncrop x {}, crop@dyncrop y {}",
                    seg.start, seg.crop.width, seg.crop.height, seg.crop.x, seg.crop.y
                )
            })
            .collect::<Vec<_>>()
            .join("; ")
    }

    /// Fallback: Render segments and concatenate with fixed timestamps.
    ///
    /// Uses setpts=PTS-STARTPTS on each segment to ensure clean concatenation.
    async fn render_and_concat_segments_fixed<P: AsRef<Path>>(
        &self,
        input: P,
        output: P,
        segments: &[CropSegment],
        base_start: f64,
    ) -> MediaResult<()> {
        let input = input.as_ref();
        let output = output.as_ref();

        let temp_dir = TempDir::new()?;
        let mut segment_files = Vec::new();

        // Render each segment with timestamp normalization
        for (i, segment) in segments.iter().enumerate() {
            let seg_path = temp_dir.path().join(format!("segment_{:04}.mp4", i));
            self.render_single_segment_normalized(input, &seg_path, segment, base_start).await?;
            segment_files.push(seg_path);
        }

        // Concatenate with stream copy (segments are now normalized)
        let concat_list_path = temp_dir.path().join("concat.txt");
        let concat_content: String = segment_files
            .iter()
            .map(|p| format!("file '{}'\n", p.display()))
            .collect();

        tokio::fs::write(&concat_list_path, &concat_content).await?;

        let mut cmd = Command::new("ffmpeg");
        cmd.args([
            "-y",
            "-f", "concat",
            "-safe", "0",
            "-i", concat_list_path.to_str().unwrap_or(""),
            // Stream copy - segments are already encoded with normalized timestamps
            "-c:v", "copy",
            "-c:a", "copy",
            "-movflags", "+faststart",
            output.to_str().unwrap_or(""),
        ])
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

        let output_result = cmd.output().await.map_err(|e| {
            MediaError::ffmpeg_failed(format!("FFmpeg concat failed: {}", e), None, None)
        })?;

        if !output_result.status.success() {
            let stderr = String::from_utf8_lossy(&output_result.stderr);
            return Err(MediaError::ffmpeg_failed(
                "FFmpeg concat failed",
                Some(stderr.to_string()),
                output_result.status.code(),
            ));
        }

        Ok(())
    }

    /// Render a segment with PTS normalization for clean concatenation.
    async fn render_single_segment_normalized<P: AsRef<Path>>(
        &self,
        input: P,
        output: P,
        segment: &CropSegment,
        base_start: f64,
    ) -> MediaResult<()> {
        let input = input.as_ref();
        let output = output.as_ref();
        let crop = &segment.crop;

        let start = base_start + segment.start;
        let duration = segment.end - segment.start;

        // Key fix: setpts=PTS-STARTPTS normalizes timestamps
        // Scale to exact 1080×1920 portrait output with square pixels
        let vf = format!(
            "crop={}:{}:{}:{},setpts=PTS-STARTPTS,scale={}:{}:flags=lanczos,setsar=1",
            crop.width, crop.height, crop.x, crop.y,
            PORTRAIT_WIDTH, PORTRAIT_HEIGHT
        );

        let mut cmd = Command::new("ffmpeg");
        cmd.args([
            "-y",
            "-ss", &format!("{:.3}", start),
            "-i", input.to_str().unwrap_or(""),
            "-t", &format!("{:.3}", duration),
            "-vf", &vf,
            // Audio timestamp normalization
            "-af", "aresample=async=1:first_pts=0",
            "-c:v", "libx264",
            "-preset", &self.config.render_preset,
            "-crf", &self.config.render_crf.to_string(),
            "-c:a", "aac",
            "-b:a", "128k",
            "-pix_fmt", "yuv420p",
            output.to_str().unwrap_or(""),
        ])
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

        let output_result = cmd.output().await.map_err(|e| {
            MediaError::ffmpeg_failed(format!("Failed to render segment: {}", e), None, None)
        })?;

        if !output_result.status.success() {
            let stderr = String::from_utf8_lossy(&output_result.stderr);
            return Err(MediaError::ffmpeg_failed(
                "Segment render failed",
                Some(stderr.to_string()),
                output_result.status.code(),
            ));
        }

        Ok(())
    }

    /// Group crop windows into segments with similar crops.
    fn group_segments(&self, windows: &[CropWindow]) -> Vec<CropSegment> {
        if windows.is_empty() {
            return Vec::new();
        }

        let tolerance = 0.1; // 10% tolerance
        let mut segments = Vec::new();
        let mut current_start = windows[0].time;
        let mut current_crop = windows[0];

        for window in windows.iter().skip(1) {
            if self.crops_differ(&current_crop, window, tolerance) {
                // End current segment
                segments.push(CropSegment {
                    start: current_start,
                    end: window.time,
                    crop: current_crop,
                });
                current_start = window.time;
                current_crop = *window;
            }
        }

        // Add final segment
        segments.push(CropSegment {
            start: current_start,
            end: windows.last().unwrap().time + 1.0, // Extend slightly
            crop: current_crop,
        });

        segments
    }

    /// Check if two crops are significantly different.
    fn crops_differ(&self, crop1: &CropWindow, crop2: &CropWindow, tolerance: f64) -> bool {
        let threshold = (crop1.width as f64 * tolerance) as i32;
        (crop1.x - crop2.x).abs() > threshold
            || (crop1.y - crop2.y).abs() > threshold
            || (crop1.width - crop2.width).abs() > threshold
    }

    /// Render a single segment.
    async fn render_single_segment<P: AsRef<Path>>(
        &self,
        input: P,
        output: P,
        segment: &CropSegment,
        base_start: f64,
    ) -> MediaResult<()> {
        let input = input.as_ref();
        let output = output.as_ref();
        let crop = &segment.crop;

        let start = base_start + segment.start;
        let duration = segment.end - segment.start;

        // Crop to 9:16 region, then scale to exact 1080×1920 portrait output
        let vf = format!(
            "crop={}:{}:{}:{},scale={}:{}:flags=lanczos,setsar=1",
            crop.width, crop.height, crop.x, crop.y,
            PORTRAIT_WIDTH, PORTRAIT_HEIGHT
        );

        let mut cmd = Command::new("ffmpeg");
        cmd.args([
            "-y",
            "-ss",
            &format!("{:.3}", start),
            "-i",
            input.to_str().unwrap_or(""),
            "-t",
            &format!("{:.3}", duration),
            "-vf",
            &vf,
            "-c:v",
            "libx264",
            "-preset",
            &self.config.render_preset,
            "-crf",
            &self.config.render_crf.to_string(),
            "-c:a",
            "aac",
            "-b:a",
            "128k",
            "-movflags",
            "+faststart",
            output.to_str().unwrap_or(""),
        ])
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

        let output_result = cmd.output().await.map_err(|e| {
            MediaError::ffmpeg_failed(format!("Failed to run FFmpeg: {}", e), None, None)
        })?;

        if !output_result.status.success() {
            let stderr = String::from_utf8_lossy(&output_result.stderr);
            return Err(MediaError::ffmpeg_failed(
                "FFmpeg segment render failed",
                Some(stderr.to_string()),
                output_result.status.code(),
            ));
        }

        Ok(())
    }

    /// Compute median crop from windows.
    fn compute_median_crop(&self, windows: &[CropWindow]) -> CropWindow {
        if windows.is_empty() {
            return CropWindow::new(0.0, 0, 0, PORTRAIT_WIDTH as i32, PORTRAIT_HEIGHT as i32);
        }

        let mut x_vals: Vec<i32> = windows.iter().map(|w| w.x).collect();
        let mut y_vals: Vec<i32> = windows.iter().map(|w| w.y).collect();
        let mut w_vals: Vec<i32> = windows.iter().map(|w| w.width).collect();
        let mut h_vals: Vec<i32> = windows.iter().map(|w| w.height).collect();

        x_vals.sort();
        y_vals.sort();
        w_vals.sort();
        h_vals.sort();

        let mid = windows.len() / 2;

        CropWindow::new(
            windows[mid].time,
            x_vals[mid],
            y_vals[mid],
            w_vals[mid],
            h_vals[mid],
        )
    }
}

/// A crop segment for rendering.
#[derive(Debug, Clone)]
struct CropSegment {
    /// Start time relative to clip start
    start: f64,
    /// End time relative to clip start
    end: f64,
    /// Crop window for this segment
    crop: CropWindow,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_group_segments_static() {
        let config = IntelligentCropConfig::default();
        let renderer = IntelligentRenderer::new(config);

        let windows = vec![
            CropWindow::new(0.0, 100, 100, 500, 500),
            CropWindow::new(1.0, 102, 101, 500, 500),
            CropWindow::new(2.0, 101, 100, 500, 500),
        ];

        let segments = renderer.group_segments(&windows);
        assert_eq!(segments.len(), 1); // Should be one segment
    }

    #[test]
    fn test_group_segments_dynamic() {
        let config = IntelligentCropConfig::default();
        let renderer = IntelligentRenderer::new(config);

        let windows = vec![
            CropWindow::new(0.0, 100, 100, 500, 500),
            CropWindow::new(1.0, 100, 100, 500, 500),
            CropWindow::new(2.0, 300, 300, 500, 500), // Big change
            CropWindow::new(3.0, 300, 300, 500, 500),
        ];

        let segments = renderer.group_segments(&windows);
        assert_eq!(segments.len(), 2); // Should be two segments
    }

    #[test]
    fn test_median_crop() {
        let config = IntelligentCropConfig::default();
        let renderer = IntelligentRenderer::new(config);

        let windows = vec![
            CropWindow::new(0.0, 100, 100, 500, 500),
            CropWindow::new(1.0, 200, 200, 600, 600),
            CropWindow::new(2.0, 150, 150, 550, 550),
        ];

        let median = renderer.compute_median_crop(&windows);
        assert_eq!(median.x, 150);
        assert_eq!(median.y, 150);
        assert_eq!(median.width, 550);
    }
}
